File: tests/patterns/julia/implicit_return.jl
# ERROR:
f() = "implicit return for function definition without function keyword"

# ERROR:
() -> "implicit return for lambda"

function f()
  nothing
  # ERROR:
  "implicit return last statement"
end

function f(x)
  if x < 0
    # ERROR:
    "implicit return if"
  elseif x == 0
    # ERROR:
    "implicit return elsif"
  else
    # ERROR:
    "implicit return else"
  end
end

function f(x)
  if x == 0
    "no implicit return because not the last statement"
  else
    "no implicit return because not the last statement"
  end
  # ERROR:
  "implicit return last statement after non trivial"
end

function f()
  function f()
    nil
    # ERROR:
    "implicit return nested function even when function is not the last statement"
  end
  # ERROR:
  "implicit return last statement after nested function"
end

function f()
  try
    # ERROR:
    "implicit return in try clause"
  catch e
    "no implicit return in catch clause if there is no exception"
  end
end

function f()
  try
    "no implicit return in try clause if else clause is present"
  catch e
    "no implicit return in catch clause if there is no exception"
  else
    # ERROR:
    "implicit return in else clause if there is no exception in try clause"
  end
end

function f()
  try
    "no implicit return in try clause if ensure is present"
  catch e
    "no implicit return in catch clause if ensure is present"
  finally
    # ERROR:
    "implicit return in ensure clause"
  end
end

function f()
  try
    "no implicit return in try clause if ensure is present"
  catch e
    "no implicit return in catch clause if ensure is present"
  else
    "no implicit return in else clause if ensure is present"
  finally
    # ERROR:
    "implicit return in ensure clause"
  end
end

module M
  function f()
    # ERROR:
    "implicit return function inside module"
  end
end

module M
  function f(implicit_return_parameter)
    # ERROR:
    implicit_return_parameter
  end

  function g()
    # ERROR:
    f("implicit return function call")
  end
end

function f()
  # ERROR:
  return "explicit return should still work"
end

File: tests/patterns/scala/implicit_return.scala
// ERROR:
def f() = "implicit return for function definition without function keyword"

// ERROR:
val _ = () => "implicit return for lambda"

def f() =
  val ignore_me = 0
  // ERROR:
  "implicit return last statement"

def f(x) =
  if x < 0 then
    // ERROR:
    "implicit return if"
  else if x == 0 then
    // TODO:
    "implicit return else if"
  else
    // TODO:
    "implicit return else"

def f(x) =
  val _ =
  if x == 0 then
    "no implicit return because not the last statement"
  else
    "no implicit return because not the last statement"
   
  // ERROR:
  "implicit return last statement after non trivial"

def f() =
  def f() =
    // ERROR:
    "implicit return nested function even when function is not the last statement"

  // ERROR:
  "implicit return last statement after nested function"

def f() =
  try
    // TODO:
    "implicit return in try clause"
  catch
    case e => "no implicit return in catch clause if there is no exception"

def f() =
  try
    "no implicit return in try clause if finally is present"
  catch
    case e => "no implicit return in catch clause if finally is present"
  finally
    // TODO:
    "implicit return in finally clause"

package p:
  def f() =
    // ERROR:
    "implicit return function inside package"

package p:
  def f(implicit_return_parameter) =
    // ERROR:
    implicit_return_parameter

  def g() =
    // ERROR:
    f("implicit return function call")

def f() =
  // ERROR:
  return "explicit return should still work"

